openedx/features/course_duration_limits/tests/test_models.py

```diff
@@ -0,0 +1,177 @@
+            existing_enrollment.save()
+        else:
+            existing_enrollment = None
+            if enrolled_before_enabled:
+                today = config.enabled_as_of - timedelta(days=1)
+            else:
+                today = config.enabled_as_of + timedelta(days=1)
+
+        if pass_enrollment:
+            enrollment = existing_enrollment
+            user = None
+            course_key = None
+        else:
+            enrollment = None
+            user = self.user
+            course_key = self.course_overview.id
+
+        if pass_enrollment:
+            query_count = 4
+        else:
+            query_count = 5
+
+        with self.assertNumQueries(query_count):
+            enabled = CourseDurationLimitConfig.enabled_for_enrollment(
+                enrollment=enrollment,
+                user=user,
+                course_key=course_key,
+                today=today,
+            )
+            self.assertEqual(not enrolled_before_enabled, enabled)
+
+    def test_enabled_for_enrollment_failure(self):
+        with self.assertRaises(ValueError):
+            CourseDurationLimitConfig.enabled_for_enrollment(None, None, None)
+        with self.assertRaises(ValueError):
+            CourseDurationLimitConfig.enabled_for_enrollment(Mock(name='enrollment'), Mock(name='user'), None)
+        with self.assertRaises(ValueError):
+            CourseDurationLimitConfig.enabled_for_enrollment(Mock(name='enrollment'), None, Mock(name='course_key'))
+
+    @override_waffle_flag(CONTENT_TYPE_GATING_FLAG, True)
+    def test_enabled_for_enrollment_flag_override(self):
+        self.assertTrue(CourseDurationLimitConfig.enabled_for_enrollment(None, None, None))
+        self.assertTrue(CourseDurationLimitConfig.enabled_for_enrollment(Mock(name='enrollment'), Mock(name='user'), None))
+        self.assertTrue(CourseDurationLimitConfig.enabled_for_enrollment(Mock(name='enrollment'), None, Mock(name='course_key')))
+
+    @ddt.data(True, False)
+    def test_enabled_for_course(
+        self,
+        before_enabled,
+    ):
+        config = CourseDurationLimitConfig.objects.create(
+            enabled=True,
+            course=self.course_overview,
+            enabled_as_of=date.today(),
+        )
+
+        if before_enabled:
+            today = config.enabled_as_of - timedelta(days=1)
+        else:
+            today = config.enabled_as_of + timedelta(days=1)
+
+        course_key = self.course_overview.id
+
+        self.assertEqual(
+            not before_enabled,
+            CourseDurationLimitConfig.enabled_for_course(
+                course_key=course_key,
+                today=today,
+            )
+        )
+
+    @ddt.data(
+        # Generate all combinations of setting each configuration level to True/False/None
+        *itertools.product(*[(True, False, None)]*4)
+    )
+    @ddt.unpack
+    def test_config_overrides(self, global_setting, site_setting, org_setting, course_setting):
```

Comment:
perhaps if we have a general docstring about how these function we could also copy paste it here to help understand the test - since we're using ddt to create all the test cases it is less obvious just looking at the test what the expected results for a given case are